src/harness/sas_objects.py

```diff
@@ -0,0 +1,387 @@
+            grant_object.setState(Grant.OUT_OF_SYNC)
+        elif (heartbeat_response_code == 500):
+            grant_object.setState(Grant.TERMINATED)
+        elif (heartbeat_response_code == 501):
+            grant_object.setState(Grant.SUSPENDED)
+        else:
+            assert(not "Invalid grant response code")
+
+
+    def performHeartbeatAndUpdateGrants(self):
+        """
+            This function construct heartbeat request for all active grants and updates the grants
+        """
+        for cbsd_object_item in self.cbsd_objects.values():
+            heartbeat_responses_with_operation_parameters = {}
+
+            #Heartbeat on all active grants
+            heartbeat_requests = cbsd_object_item.constructHeartbeatRequestForAllActiveGrants()
+            for heartbeat_request in heartbeat_requests:
+                heartbeat_response = self.testcase._sas.Heartbeat(
+                    heartbeat_request, self.ssl_cert, self.ssl_key)['heartbeatResponse'][0]
+                grant_object = cbsd_object_item.grant_objects[heartbeat_response['grantId']]
+                if heartbeat_response['response']['responseCode'] == 0:
+                    transmit_expire_time = datetime.strptime(heartbeat_response['transmitExpireTime'],
+                                                             '%Y-%m-%dT%H:%M:%SZ')
+                    self.testcase.assertLess(datetime.utcnow(), transmit_expire_time)
+                    self.testcase.assertLessEqual(
+                        (transmit_expire_time - datetime.utcnow()).total_seconds(), 240)
+                    grant_object.setState(Grant.AUTHORIZED)
+                else:
+                    self.mapResponseCodeToGrantState(
+                        heartbeat_response['response']['responseCode'], grant_object)
+
+                    #If any heartbeats contain suggested operating params,
+                    #relinquish the corresponding grant
+                    if 'operationParam' in heartbeat_response:
+                        relinquish_request = {
+                            'relinquishmentRequest': [{
+                                'cbsdId': grant_object.getGrantRequest()['cbsdId'],
+                                'grantId': grant_object.getGrantRequest()['grantId']
+                            }]
+                        }
+                        self.testcase._sas.Relinquishment(
+                            relinquish_request, self.ssl_cert, self.ssl_key)
+                        heartbeat_responses_with_operation_parameters.update(
+                            {heartbeat_response['cbsdId']: heartbeat_response})
+                        del cbsd_object_item.grant_objects[heartbeat_response['grant_id']]
+
+            #For all grants in the previous step, request a new grant with the
+            #suggested operation parameters
+            for cbsd_id, heartbeatResponse in heartbeat_responses_with_operation_parameters.items():
+                grant_0 = json.load(
```

Comment:
No need to load, since you're overwriting all of the values anyway.

Reply:
changed